parasolr/query.py

```diff
@@ -161,67 +167,52 @@ def get_facets(self) -> Dict[str, int]:
-        double_underscore_regex = re.compile(r'_{2}(?=[A-Za-z]+$)')
-
-        # value is a list, join with OR logic for all values in list,
-        # include the empty value for a missing/empty field
-        if key.endswith('__in'):
-            # filter out a negative search and set flag
-            null_filter = bool([item for item in value if not item])
-
-            value = [item for item in value if item]
-            key = re.split(double_underscore_regex, key)[0]
+        # or __range=(start, end)
+
+        # NOTE: Assuming there is only one LOOKUP_SEP without error handling
+        key, lookup = split_key
+
+        if lookup == 'in':
+            # value is a list, join with OR logic for all values in list,
+            # treat '' or None values as flagging an exists query
+            implicit_exists = False
+            if '' in value or None in value:
+                implicit_exists = True
+            value = filter(lambda x: x not in ['', None], value)
+            _filter = '%s:(%s)' % (key, ' OR '.join(value))
+            if not implicit_exists:
+                return _filter
+            else:
+                # This query handles the fact that query syntax does not
+                # support the simpler positive case. Instead, we do a
+                # negative lookup that negates a positive lookup for
+                # all possible values and double-negates a lookup
+                # for any filtered values (thus producing a positive)
+                # The final output is something like:
+                # -(item_type:[* TO *] OR item_type: book OR periodical)
+                return '-(%s:%s OR -%s)' % (key, ALL_VALUES, _filter)
+        if lookup == 'exists':
+            # Look for all possible values, and either negatve or not,
+            # depending on the boolean of value.
             if value:
-                _filter = '(%s)' % ' OR '.join(value)
-
-        # dictionary for easier interpolation of values
-        filter_vals = {
-            'filter': _filter if _filter else value,
-            'key': key
-        }
-
-        # if filter_vals['filter] can be cast to int,
-        # do so for comparison against 0 since we don't want to treat
-        # that value as implicitly falsy
-        try:
-            filter_vals['filter'] = int(filter_vals['filter'])
-        except ValueError:
-            pass
-
-        if not filter_vals['filter'] and filter_vals['filter'] != 0:
-            null_filter = True
-
-        # there is a key:value pair and a search for a null via an empty string
-        # or None
-        if null_filter and value:
-            return '-(%(key)s:[* TO *] -%(key)s:%(filter)s)' % filter_vals
-        # there is a search for a null field
-        elif null_filter and not value:
-            return '-%(key)s:[* TO *]' % filter_vals
-        # there is a simple key/value pair
-        return '%(key)s:%(filter)s' % filter_vals
-
+                return '%s:%s' % (key, ALL_VALUES)
```

Comment:
Recommend combining these two cases; either generate and add the `-` depending on the value, or even include it in the `%` string (`'' if value else "-"`).